### PR TITLE
[FEATURE] Change array cache to HTAB cache for wal_block_numbers

### PR DESCRIPTION
**Pull request by var77**
## Issue
https://github.com/lanterndata/lanterndb/issues/21

## Description
Changed `wal_retriever_block_numbers` array to Postgres `HTAB` (hash table) for caching block numbers.

### CODE REVIEW

File: test/expected/hnsw_insert_array.out [hunk 1]
```diff
@@ -147,9 +161,9 @@ SELECT id, ROUND((v <-> '{1,0,0,0,0,0,21,35,1,0,0,0,0,77,51,42,66,2,0,0,0,86,140
   39 | 56265.00
  886 | 71637.00
  402 | 77013.00
- 518 | 85024.00
  340 | 87261.00
  331 | 87796.00
+ 518 | 85024.00
```
Comment (davkhech) [inline test/expected/hnsw_insert_array.out:166]: is this fine?

File: src/hnsw/external_index.c [hunk 2]
```diff
@@ -625,17 +623,11 @@ static inline void *wal_index_node_retriever_exact(int id)
...
-        // todo:: verify that this sizeof is evaled compile time and blockmap_page is not derefed
-        //  in any way, probably worth changing it. looks strange to -> on an uninitialized pointer
-        int write_size = sizeof(blockmap_page->blocknos);
-        // if this is th elast page and blocknos is not filled up, only read the part that is filled up
-        if(id_offset + HNSW_BLOCKMAP_BLOCKS_PER_PAGE > HEADER_FOR_EXTERNAL_RETRIEVER.num_vectors) {
-            write_size = (HEADER_FOR_EXTERNAL_RETRIEVER.num_vectors - id_offset) * sizeof(blockmap_page->blocknos[ 0 ]);
-        }
         buf = ReadBufferExtended(INDEX_RELATION_FOR_RETRIEVER, MAIN_FORKNUM, blockmapno, RBM_NORMAL, NULL);
         // you might expect that this is unnecessary, since we always unlock the pagemap page right after reading the
         // necessary information into wal_retriever_block_numbers
```
Comment (Ngalstyan4) [inline src/hnsw/external_index.c:633]: change to `wal_retriever_block_numbers_cache`

File: src/hnsw/external_index.c [hunk 3]
```diff
@@ -176,7 +179,7 @@ void StoreExternalIndex(Relation        index,
     wal_retriever_block_numbers = palloc0(sizeof(BlockNumber) * num_added_vectors);
```
Comment (Ngalstyan4) [inline src/hnsw/external_index.c:179]: should this be `pfree`ed at the end of the function?

File: src/hnsw/external_index.c [hunk 4]
```diff
@@ -7,15 +7,16 @@
...
 #include "common/relpath.h"
 #include "insert.h"
 #include "pg_config.h"       // BLCKSZ
 #include "storage/bufmgr.h"  // Buffer
 #include "usearch.h"
 #include "utils/relcache.h"
 
-static inline void  fill_blockno_mapping();
-static BlockNumber *wal_retriever_block_numbers = NULL;
+static inline void fill_blockno_mapping();
```
Comment (Ngalstyan4) [inline src/hnsw/external_index.c:18]: declaration mismatch? the later version has an argument.

File: src/hnsw/cache.c [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+{
+    CacheEntry *entry;
+
+    entry = (CacheEntry *)hash_search(cache, key, HASH_ENTER, NULL);
+    entry->value = blockno;
+}
+
+void cache_destroy(HTAB *cache)
+{
+    // MemoryContextDelete(cache->hctl.hcxt);
```
Comment (Ngalstyan4) [inline src/hnsw/cache.c:46]: This should not be commented out right? We created the context here, so it should be safe to also destroy it

**Review by Ngalstyan4 — COMMENTED** [review]
Looks good! We should stress-test and benchmark this next

File: test/expected/hnsw_insert_array.out (hunk 1 above)
Comment (Ngalstyan4) [reply test/expected/hnsw_insert_array.out:166]: I think we overlooked an issue in a previous PR (#18) where the regression tests for `real[]` column typed tables did not actually create an index.
It seems Varik fixed it there, which caused some of the changes.

Without the index we do exact ordering, which is always 100% accurate. With the index, we do approximate ordering so anomalies like the above are acceptible.

File: test/expected/hnsw_insert_array.out [hunk 6]
```diff
@@ -26,9 +26,11 @@ psql:test/sql/hnsw_insert_array.sql:6: INFO:  done init usearch index
 psql:test/sql/hnsw_insert_array.sql:6: INFO:  inserted 8 elements
 psql:test/sql/hnsw_insert_array.sql:6: INFO:  done saving 8 vectors
 CREATE INDEX
-CREATE INDEX ON sift_base1k USING hnsw (v);
+CREATE INDEX ON sift_base1k USING hnsw (v) WITH (dims=128);
 psql:test/sql/hnsw_insert_array.sql:7: INFO:  done init usearch index
-psql:test/sql/hnsw_insert_array.sql:7: ERROR:  Wrong number of dimensions: 128 instead of 3 expected
```
[the remark this answers is not in the document]
Comment (var77) [reply test/expected/hnsw_insert_array.out:31]: Yes there's a default dimension of 3, and also there are tests to check the size checking errors https://github.com/lanterndata/lanterndb/blob/main/test/sql/hnsw_type_checks.sql

File: src/hnsw/external_index.c (hunk 3 above)
Comment (var77) [reply src/hnsw/external_index.c:179]: Yes it is being `pfree` ed at the end of  the function [here](https://github.com/lanterndata/lanterndb/pull/26/files/67473f87e37e34709ba80f40bf99a7106d803fe6#diff-37df803808fb881daf0c4bda9979f2cb5ac24dc1f1ab34ca668ad2184dfc96eaR253)

**Pull request merged**